presto-main/src/main/java/io/prestosql/server/security/UserExtraction.java

```diff
@@ -0,0 +1,124 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.prestosql.server.security;
+
+import com.fasterxml.jackson.annotation.JsonProperty;
+import com.google.common.collect.ImmutableList;
+import io.airlift.json.JsonCodec;
+
+import javax.annotation.Nullable;
+
+import java.io.File;
+import java.io.IOException;
+import java.nio.file.Files;
+import java.util.List;
+import java.util.Optional;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+import static com.google.common.base.MoreObjects.firstNonNull;
+import static com.google.common.base.Preconditions.checkArgument;
+import static java.lang.Boolean.TRUE;
+import static java.util.Objects.requireNonNull;
+
+final class UserExtraction
+{
+    private static final JsonCodec<UserExtractionRules> RULES_JSON_CODEC = JsonCodec.jsonCodec(UserExtractionRules.class);
+    private final List<Rule> rules;
+
+    public static UserExtraction createUserExtraction(@Nullable String userExtractionPattern, @Nullable File userExtractionFile)
+            throws IOException
+    {
+        if (userExtractionPattern != null) {
+            checkArgument(userExtractionFile == null, "user extraction pattern and file can not both be set");
+            return new UserExtraction(userExtractionPattern);
+        }
+        if (userExtractionFile != null) {
+            return new UserExtraction(userExtractionFile);
+        }
+        return new UserExtraction("(.*)");
+    }
+
+    private UserExtraction(String pattern)
+    {
+        rules = ImmutableList.of(new Rule(pattern, "$1", true));
+    }
+
+    private UserExtraction(File rulesFile)
+            throws IOException
+    {
+        requireNonNull(rulesFile, "rulesFile is null");
+
+        rules = RULES_JSON_CODEC.fromJson(Files.readAllBytes(rulesFile.toPath())).getRules();
+    }
+
+    public String extractUser(String principal)
```

Comment:
I guess it should return `Optional<String>` because I image that not always it is possible to extract user from principal. Like how to extract user from token